// ss-message/lib/plugins/base.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _botLang = require('bot-lang');

var _botLang2 = _interopRequireDefault(_botLang);

var _nlp_compromise = require('nlp_compromise');

var _nlp_compromise2 = _interopRequireDefault(_nlp_compromise);

var _lemmer = require('lemmer');

var _lemmer2 = _interopRequireDefault(_lemmer);

var _async = require('async');

var _async2 = _interopRequireDefault(_async);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _pos = require('pos');

var _pos2 = _interopRequireDefault(_pos);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const tagger = new _pos2.default.Tagger();
const lexer = new _pos2.default.Lexer();

const addTags = function addTags(cb) {
  this.message.tags = _botLang2.default.tag.all(this.message.original);
  cb();
};

const addNlp = function addNlp(cb) {
  this.message.nlp = (0, _nlp_compromise2.default)(this.message.original);
  cb();
};

const addEntities = function addEntities(cb) {
  const entities = this.message.nlp.match('(#Person|#Place|#Organization)').asArray();
  this.message.entities = entities;

  // For legacy support
  this.message.names = this.message.nlp.people().asArray();
  cb();
};

const addDates = function addDates(cb) {
  this.message.dates = this.message.nlp.dates().asArray();
  cb();
};

const addWords = function addWords(cb) {
  this.message.words = this.message.clean.split(' ');
  cb();
};

const addPos = function addPos(cb) {
  this.message.nouns = this.message.nlp.match('#Noun').asArray();
  this.message.adverbs = this.message.nlp.match('#Adverb').asArray();
  this.message.verbs = this.message.nlp.match('#Verb').asArray();
  this.message.adjectives = this.message.nlp.match('#Adjective').asArray();
  this.message.pronouns = this.message.nlp.match('#Pronoun').asArray();

  // Fix for pronouns getting mixed in with nouns
  _lodash2.default.pullAll(this.message.nouns, this.message.pronouns);
  cb();
};

const hasExpression = function hasExpression(cb) {
  const expressionTerms = ["add", "plus", "and", "+", "-", "minus", "subtract", "x", "times", "multiply", "multiplied", "of", "divide", "divided", "/", "half", "percent", "%"];

  const containsArithmeticTerm = _lodash2.default.some(this.message.words, word => {
    return expressionTerms.indexOf(word) !== -1;
  });

  const hasTwoNumbers = this.message.nlp.match('#Value').asArray().length >= 2;
  this.message.expression = containsArithmeticTerm && hasTwoNumbers;
  cb();
};

const pennToWordnet = function pennToWordnet(pennTag) {
  if (pennTag[0] === 'J') {
    return 'a';
  } else if (pennTag[0] === 'V') {
    return 'v';
  } else if (pennTag[0] === 'N') {
    return 'n';
  } else if (pennTag[0] === 'R') {
    return 'r';
  }
  return null;
};

const fixup = function fixup(cb) {
  // fix numeric forms
  // twenty-one => 21
  this.message.clean = (0, _nlp_compromise2.default)(this.message.clean).values().toNumber().plaintext();

  // singalize / lemmatize
  // This does a slightly better job than `.split(" ")`
  this.message.words = lexer.lex(this.message.clean);
  const taggedWords = tagger.tag(this.message.words);

  const itor = (hash, next) => {
    const word = hash[0].toLowerCase();
    const tag = pennToWordnet(hash[1]);

    if (tag) {
      return _lemmer2.default.lemmatize(`${ word }#${ tag }`, next);
    }
    // Some words don't have a tag ie: like, to.
    return next(null, [word]);
  };

  _async2.default.map(taggedWords, itor, (err, transformed) => {
    this.message.lemString = _lodash2.default.map(_lodash2.default.flatten(transformed), a => a.split('#')[0]).join(' ');
    cb();
  });
};

const addQuestionTypes = function addQuestionTypes(cb) {
  // Classify Question
  const questionWords = ['who', 'whose', 'whom', 'what', 'where', 'when', 'why', 'which', 'name', 'did', 'do', 'does', 'have', 'had', 'has'];
  let isQuestion = false;

  if (this.message.raw.slice(-1) === '?') isQuestion = true;
  if (questionWords.indexOf(this.message.words[0]) !== -1) isQuestion = true;
  this.message.isQuestion = isQuestion;

  cb();
};

// Order here matters
exports.default = {
  addTags,
  addNlp,
  addEntities,
  addDates,
  addPos,
  addWords,
  hasExpression,
  fixup,
  addQuestionTypes
};
